src/common/openssl_opts_handler.cc

```diff
@@ -0,0 +1,134 @@
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ */
+
+#include "openssl_opts_handler.h"
+
+#include <openssl/bio.h>
+#include <openssl/conf.h>
+#include <openssl/engine.h>
+#include <mutex>
+#include <vector>
+#include <algorithm>
+
+#include "common/debug.h"
+#include "global/global_context.h"
+#include "include/str_list.h"
+
+// -----------------------------------------------------------------------------
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_common
+#undef dout_prefix
+#define dout_prefix _prefix(_dout)
+
+static ostream &_prefix(std::ostream *_dout)
+{
+  return *_dout << "OpenSSLOptsHandler: ";
+}
+// -----------------------------------------------------------------------------
+
+string construct_engine_conf(const string &opts)
+{
+  const string conf_header = "openssl_conf=openssl_def\n[openssl_def]\n";
+  const string engine_header = "engines=engine_section\n[engine_section]\n";
+
+  string engine_id, engine_statement, engine_detail;
+  const string id_prefix = "engine";
+  const string suffix = "_section";
+  const char delimiter = '\n';
+
+  int index = 1;
+  vector<string> confs = get_str_vec(opts, ":");
+  for (auto conf : confs) {
+    // Construct engine section statement like "engine1=engine1_section"
+    engine_id = id_prefix + to_string(index++);
+    engine_statement += engine_id + "=" + engine_id + suffix + delimiter;
+
+    // Adapt to OpenSSL parser
+    // Replace ',' with '\n' and add section in front
+    replace(conf.begin(), conf.end(), ',', delimiter);
+    engine_detail += "[" + engine_id + suffix + "]" + delimiter;
+    engine_detail += conf + delimiter;
+  }
+
+  return conf_header + engine_header + engine_statement + engine_detail;
+}
+
+void do_free(BIO *mem, CONF *conf)
+{
+  BIO_free(mem);
+  NCONF_free(conf);
+}
+
+string get_openssl_error()
+{
+  BIO *bio = BIO_new(BIO_s_mem());
+  if (bio == nullptr) {
+    return "failed to create bio for error printing";
+  }
+  ERR_print_errors(bio);
+  char* buf;
+  size_t len = BIO_get_mem_data(bio, &buf);
+  string ret(buf, len);
+  BIO_free(bio);
+  return ret;
+}
+
+void load_module(const string &engine_conf)
+{
+  BIO *mem = BIO_new_mem_buf(engine_conf.c_str(), engine_conf.size());
+  if (mem == nullptr) {
+    derr << "failed to new memory" << dendl;
+    return;
+  }
+
+  CONF *conf = NCONF_new(nullptr);
+  if (conf == nullptr) {
+    derr << "failed to new conf" << dendl;
+    BIO_free(mem);
+    return;
+  }
+
+  if (NCONF_load_bio(conf, mem, nullptr) <= 0) {
+    derr << "failed to load bio: " << get_openssl_error() << dendl;
+    do_free(mem, conf);
+    return;
+  }
+
+  OPENSSL_load_builtin_modules();
+  ENGINE_load_builtin_engines();
+
+  if (CONF_modules_load(
+          conf, nullptr,
+          CONF_MFLAGS_DEFAULT_SECTION | CONF_MFLAGS_IGNORE_MISSING_FILE) <= 0) {
+    derr << "failed to load modules: " << get_openssl_error() << dendl;
```

Comment:
What's the error policy? Is it enough to print a message and continue? Have we considered asserting?

Reply:
when failed to load one engine ( and if there were engines left,  they would not be loaded), the process would still be able to call openssl algorithm, like MD5/AES, although the performance may be not as good as using engine. so maybe we would like to keep the process running when this happened?

Reply:
in my test, it looks that "an informative error message" as u mentioned has been printed and logged by using `derr`.  did i misunderstand somethine here?

Reply:
Well, just `failed to new conf` doesn't seem to be so informative, to be honest. I meant rather something like explicitly saying that the error is related with OpenSSL and, as the engine loading is affected, e.g. the intended acceleration won't be available.

Reply:
i see, i'll fix it.

Reply:
@rzarzynski , I've made changes. please check.